tests/store/artifact/test_ftp_artifact_repo.py

```diff
@@ -211,32 +211,93 @@ def test_log_artifact_multiple_calls(ftp_mock, tmpdir):
     assert ftp_mock.storbinary.call_args_list[0][0][0] == "STOR test2.txt"
 
 
+def __posixpath_parents(pathname, root):
+    parents = [posixpath.dirname(pathname)]
+    root = posixpath.normpath(root)
+    while parents[-1] != "/" and parents[-1] != root:
+        parents.append(posixpath.dirname(parents[-1]))
+    return parents
```

Comment:
correct me if I'm wrong, but I think this function assumes:

- `pathname` is an abosolute path
-  `pathname` startswith `root`

Can we add verify these conditions with `assert` to prevent this function from falling into an infinite loop?

Reply:
```suggestion
def __posixpath_parents(pathname, root):
    assert posixpath.isabs(pathname)
    assert pathname.startswith(root)
    
    parents = [posixpath.dirname(pathname)]
    root = posixpath.normpath(root)
    while parents[-1] != "/" and parents[-1] != root:
        parents.append(posixpath.dirname(parents[-1]))
    return parents
```